src/dune_engine/dune_project.ml

```diff
@@ -551,7 +568,96 @@ module Toggle = struct
     field_o name decode
 end
 
-let anonymous ~dir = infer ~dir Package.Name.Map.empty
+let anonymous ~dir ?info ?(packages=Package.Name.Map.empty) () = infer ~dir ?info packages
+
+let encode : t -> Dune_lang.t list =
+ fun { name
+     ; version
+     ; dune_version
+     ; info
+     ; packages
+     ; extension_args
+     ; implicit_transitive_deps
+     ; wrapped_executables
+     ; executables_implicit_empty_intf
+     ; accept_alternative_dune_file_name
+     ; generate_opam_files
+     ; use_standard_c_and_cxx_flags
+     ; dialects
+     ; explicit_js_mode
+     ; format_config
+     ; strict_package_deps
+     ; cram
+     ; subst_config
+     (* Metadata that is derived from, but not represented in dune-project files *)
+     ; file_key = _
+     ; parsing_context = _
+     ; stanza_parser = _
+     ; project_file = _
+     ; root = _
+     } ->
+  let open Dune_lang.Encoder in
+  let lang = Lang.get_exn "dune" in
+  let extension_args : Dune_lang.t list =
+    let _ = extension_args in
+    (* TODO *)
+    []
```

Comment:
iiuc, proper encoding of a `dune-project` file will need to extract data from this `Univ_map.t` to synthesize the `using` stanza, but to do that I think we'll need to provide a way of converting a `Univ_map.t` into a less opaque structure (like a list of tuples?). I don't know if the right way to do this would go going through `to_dyn: t -> Dyn.t` (I can't figure out what `Dyn.t` is for, as [the interface is undocumented](https://github.com/ocaml/dune/tree/4ca86d409ef81598750defc3f782e4c90d69c475/otherlibs/stdune-unstable/dyn.mli) ) or to write a new function for the `Univ_map` type. @jeremiedimino: it looks like you've added some functions to the `Univ_map` last year. Any thoughts on this?

Btw, I haven't been able find any docs on this `using` stanza in the dune-project files : a pointer here would be helpful.

Reply:
`Dyn` stands for "dynamic values". The `Dyn.t` type is a representation of OCaml values such that they can be processed without knowing their type. Converting a `Univ_map.t` to a `(string * Dyn.t) list` seems fine, you can add the following function to the `Dyn` module for that:


```ocaml

let to_dyns (t : t) =
  Int.Map.values t
  |> List.map ~f:(fun (Binding.T (key, v)) ->
         let (module K) = key in
         (K.name, K.to_dyn v)))
```

Reply:
Thanks!

Reply:
So added that function, but then I also decided I just don't want to fuss with figuring out how to reconstruct this data after it's been split into three different collections, each of which uses different kinds of abstractions that obscure the content. 

I'd welcome additions to this PR to add the proper encoding for this data, and I'd be happy to remove the `to_dyns` function added, but I'm not going to write the encoding of for this data myself, since it is totally orthogonal  for the feature I'm responsible for adding here.